### PR TITLE
Cbf nhg

### PR DESCRIPTION
**Pull request by abanu-ms**
**What I did**
Added support for Class-Based Forwarding next hop groups.

**Why I did it**
To support CBF routes.

**How I verified it**
Unit tested it.

**Details if related**
- Created some base classes that CbfNhgOrch and NonCbfNhgOrch (the old NhgOrch) inherit
- Created a (new) common NhgOrch that will handle the requests between the two other Orchs

### CODE REVIEW

File: orchagent/nexthopgroup.h [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+     * decremented when an object is desynced.  This will also account for the
+     * groups created by RouteOrch.
+     */
+    static unsigned m_syncedCount;
+};
+
+/*
+ * NhgCommon class representing the common base class between
+ * NextHopGroup and CbfNextHopGroup classes.
+ */
```
[the remark this answers is not in the document]
Comment (abanu-ms) [reply orchagent/nexthopgroup.h:208]: I've created NhgBase separately as RouteOrch needs to work with both CbfNhg and NonCbfNhg classes. If I'd have everything in the templated class, I couldn't grab a common base class between thee two, so NhgBase serves as a "common API" for the two of them that RouteOrch needs to use.

File: orchagent/cbfnhgorch.cpp [hunk 2]
```diff
@@ -0,0 +1,1125 @@
...
+        return make_tuple(false,
+                               vector<string>(),
+                               unordered_map<uint8_t, uint8_t>());
+    }
+
+    /*
+     * Verify that the class map contains valid data. The FC should be
+     * between 0 and 63 (inclusive), and the index should be between
+     * 1 and the size of members (inclusive). Also, the FC values
+     * should be unique (the same FC can't be mapped more than once).
```
[the remark this answers is not in the document]
Comment (abanu-ms) [reply orchagent/cbfnhgorch.cpp:270]: This isn't mapping DSCP/EXP to FCs, but FCs to NHGs. 0: 1 means "map FC 0 to use NHG 1".

File: orchagent/cbfnhgorch.cpp [hunk 3]
```diff
@@ -0,0 +1,1125 @@
...
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_INFO("Updating next hop attribute of CBF NHG member %s",
+                    to_string().c_str());
+
+    if (!isSynced())
+    {
+        SWSS_LOG_ERROR("Trying to update next hop attribute of CBF NHG member "
+                        "%s that is not synced", to_string().c_str());
+        throw logic_error("Trying to update attribute of unsynced object.");
```
[the remark this answers is not in the document]
Comment (abanu-ms) [reply orchagent/cbfnhgorch.cpp:1069]: The logic for the assert was that this issue would be hit due to developer error, not input error. The code should be wrote in such manner that we never enter this method if the member is not synced. At first this used to be an assert, but remember that asserts are compiled out in release builds.

File: orchagent/nexthopgroup.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+NhgBase::~NhgBase()
+{
+    SWSS_LOG_ENTER();
+
+    if (isSynced())
+    {
+        SWSS_LOG_ERROR("Destroying next hop group with SAI ID %lu which is"
+                        " still synced.", m_id);
+        assert(false);
```
[the remark this answers is not in the document]
Comment (abanu-ms) [reply orchagent/nexthopgroup.cpp:24]: I think I did that in the beginning, but the compiler was complaining about throwing exceptions in destructor and found some resources online that one should never throw an exception in a destructor. https://stackoverflow.com/questions/130117/throwing-exceptions-out-of-a-destructor I've resorted to assert to at least capture this error in UTs at least or debug builds.

File: orchagent/cbfnhgorch.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dshammond) [reply orchagent/cbfnhgorch.cpp:1074]: OK

File: orchagent/nexthopgroup.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dshammond) [reply orchagent/nexthopgroup.cpp:24]: OK - looks like the best approach for now.

File: orchagent/nexthopgroup.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (dshammond) [reply orchagent/nexthopgroup.h:208]: So NhgBase is the effective inheritance root, and NhgCommon is basically the templated part of Cbf/NonCbfNhg.  I think what you've got is OK then.

**Pull request merged**